### PR TITLE
Move more RTBL code to its submodule

### PR DESCRIPTION
**Pull request by badlop**
Last reorganization PR, this time mostly focused on RTBL support.

There are still some tasks, problems or ideas that should be considered:
- There's a stanza error when stopping the module (see antispam_rtbl_test.erl)
- Should add tests for new features and edge cases: multiservice, MUC filtering, ...
- When receives some IQ error responses, the corresponding service is unknown
- Probably blocked_domains (and any other data gathered from remote services) should be stored in a shared storage mnesia/ets), so mod_antispam in all the vhosts can share it instead of re-request it

### CODE REVIEW

File: src/mod_antispam_rtbl.erl [hunk 1]
```diff
@@ -22,80 +22,94 @@
...
     ejabberd_router:route_iq(SubIQ, unsubscribe_result, self()).
 
--spec request_blocked_domains(binary() | none, binary(), binary()) -> ok.
-request_blocked_domains(none, _PSNode, _From) ->
-    ok;
-request_blocked_domains(RTBLHost, RTBLDomainsNode, From) ->
+-spec request_blocked_domains([rtbl_service()] | rtbl_service(), binary()) -> any().
+request_blocked_domains(Services, From) when is_list(Services) ->
+    [request_blocked_domains(Service, From) || Service <- Services];
+request_blocked_domains(#rtbl_service{host = Host, node = Node}, From) ->
```
Comment (sstrigler) [inline src/mod_antispam_rtbl.erl:78]: Since you‘re not making a differentiation here, you‘d also request „blocked domains“ from muc_bans_sha256 nodes. This might not be problem here except for the naming of things, but I haven‘t found in your code where you‘d handle the response differently either.

Since there are no tests for `mod_muc_rtbl`, have you tested this manually, that this is still working? Could you add automated tests for `muc_bans_sha256` please?

**Review by sstrigler — COMMENTED** [review]
This PR is so large, it‘s hard to do a proper review on for me right now. Merge it yourself if you feel confident (and if that‘s possible, not sure).

File: src/mod_antispam_dump.erl [hunk 2]
```diff
@@ -126,20 +125,20 @@ open_dump_file(Name) ->
...
+            ?DEBUG("Closed ~p", [Fd]);
         {error, Reason} ->
-            ?ERROR_MSG("Cannot close ~s: ~s", [Name, file:format_error(Reason)])
+            ?ERROR_MSG("Cannot close ~p: ~s", [Fd, file:format_error(Reason)])
     end.
 
--spec reopen_dump_file(file:io_device(), binary(), binary()) -> file:io_device().
-reopen_dump_file(Fd, OldDumpFile, NewDumpFile) ->
-    close_dump_file(Fd, OldDumpFile),
+-spec reopen_dump_file2(file:io_device(), binary()) -> file:io_device().
```
Comment (sstrigler) [inline src/mod_antispam_dump.erl:139]: That‘s bad naming, please make it more telling what the difference is between those two functions. Was that change even necessary? 

Anyway, you could name this one `reopen_dump_file_by_fd` or name this one `reopen_dump_file` and name what‘s `reopen_dump_file` now to `reopen_dump_file_for_host` or similiar. I think, I like the latter more.

File: src/mod_antispam_rtbl.erl [hunk 3]
```diff
@@ -114,33 +128,127 @@ parse_items(Items) ->
...
--spec service_jid(binary()) -> jid().
-service_jid(Host) ->
-    jid:make(<<>>, Host, <<?SERVICE_JID_PREFIX, (ejabberd_cluster:node_id())/binary>>).
-
 %%--------------------------------------------------------------------
-%% Hook callbacks.
+%%| Handle iq_reply
+
+%% TODO: How to know the host and node when timeout?
+%% handle_iq_reply(timeout, blocked_domains, State) ->
```
Comment (sstrigler) [inline src/mod_antispam_rtbl.erl:135]: This is important because if there‘s a dns lookup failure or the receiving host does not answer, the service will be broken.

Comment (sstrigler) [reply src/mod_antispam_rtbl.erl:135]: That said, there should be a test for that. Maybe you could add that?

File: src/mod_antispam_dump.erl (hunk 2 above)
Comment (badlop) [reply src/mod_antispam_dump.erl:139]: >  Was that change even necessary?

When stopping the module, it calls to close the dump file, for that it attempts to get the module options to know the dump path... but the module is no longer available to get the module options!

The solution I found is to not consult module options once it's getting stopped.

But now that we have prep_stop, instead of that solution, it's possible to close the dump file during prep_stop, knowing the module options as previously...

File: src/mod_antispam_rtbl.erl (hunk 1 above)
Comment (badlop) [reply src/mod_antispam_rtbl.erl:78]: This is the data flow:
- `request_blocked_domains` sends queries to all the configured RTBL services.
- `parse_blocked_domains` manages all responses equally, regardless if they are JID or SHA. This means blocked_domains contains a mix of JIDs and SHAs.
- Finally, in `handle_call({is_blocked_domain, Domain},`, if the inspected JID is not found in blocked_domains, it searches for its SHA too.

An optimization would be to store always SHA, either because it was provided or because we calculate at storage time. Would be nice to store also JID in another field when it's known. Then search directly by SHA.

**Comment by sstrigler** [comment]
Ok, I see, thanks for clarifying. I understand it seems like this is an elegant and simple way to handle things. But both sets of data have different semantics, even if it seems they do the same, ie block messages. The biggest problem with your approach is that there is a  administrative command for ejabberdctl `get_blocked_domains <vhost>` that would stop working correctly because I wouldn’t want to see any sha hashes in there. Those would be meaningless. I really think you should keep things separate here. Would make things simpler in the long run. And cleaner. On 27 Jun 2025, at 12:51, badlop ***@***.***> wrote:﻿
@badlop commented on this pull request.



In src/mod_antispam_rtbl.erl:
>      ejabberd_router:route_iq(SubIQ, unsubscribe_result, self()).
 
--spec request_blocked_domains(binary() | none, binary(), binary()) -> ok.
-request_blocked_domains(none, _PSNode, _From) ->
-    ok;
-request_blocked_domains(RTBLHost, RTBLDomainsNode, From) ->
+-spec request_blocked_domains([rtbl_service()] | rtbl_service(), binary()) -> any().
+request_blocked_domains(Services, From) when is_list(Services) ->
+    [request_blocked_domains(Service, From) || Service <- Services];
+request_blocked_domains(#rtbl_service{host = Host, node = Node}, From) ->

This is the data flow:

request_blocked_domains sends queries to all the configured RTBL services.
parse_blocked_domains manages all responses equally, regardless if they are JID or SHA. This means blocked_domains contains a mix of JIDs and SHAs.
Finally, in handle_call({is_blocked_domain, Domain},, if the inspected JID is not found in blocked_domains, it searches for its SHA too.

An optimization would be to store always SHA, either because it was provided or because we calculate at storage time. Would be nice to store also JID in another field when it's known. Then search directly by SHA.

**Comment by badlop** [comment]
> Oh wow. Sounds complicated. Is this a real world scenario or something that popped up because we’re using mod_pubsub from our tests (because we don’t have a dedicated rtbl host)?

The current tests don't cover that scenario, so they couldn't find that problem. It appears by manual testing.
